torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp

```diff
@@ -511,6 +529,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
   GRAPH_DUMP("After remove observers :", g);
 }
 
+Value* InsertQuantDeQuantHelper::updateInputValueInGraph(
+    Value* v,
+    std::shared_ptr<Graph>& g) {
+  if (remap_values_in_observer_subgraph_.count(v) == 0) {
+    auto new_value = g->block()->addInput();
+    remap_values_in_observer_subgraph_[v] = new_value;
+    new_value->copyMetadata(v);
+    return new_value;
+  } else {
+    return remap_values_in_observer_subgraph_[v];
+  }
+}
+
+void InsertQuantDeQuantHelper::cloneNodeInGraph(
+    Node* node,
+    std::shared_ptr<Graph>& g) {
+  auto* block = g->block();
+  auto value_fn = [&](Value* v) { return updateInputValueInGraph(v, g); };
+
+  auto new_node = block->appendNode(g->createClone(node, value_fn));
+  for (size_t i = 0; i < node->outputs().size(); ++i) {
+    auto oo = node->outputs()[i];
+    auto no = new_node->outputs()[i];
+    remap_values_in_observer_subgraph_[oo] = no;
+  }
+}
+void InsertQuantDeQuantHelper::buildObserverSubgraph(
+    const std::vector<Node*>& weight_subgraph,
+    std::shared_ptr<Graph> dest_graph) {
+  // Build weight subgraph
+  for (auto n : weight_subgraph) {
+    cloneNodeInGraph(n, dest_graph);
+  }
+  LintGraph(dest_graph);
+  // Add last node output value as subgraph output.
+  dest_graph->registerOutput(
+      updateInputValueInGraph(weight_subgraph.back()->output(0), dest_graph));
+  GRAPH_DUMP("New weight observer subgraph: ", dest_graph);
+}
+
+void updateCurrInputSet(
+    const std::vector<Value*>& inputs,
+    std::unordered_set<Value*>& curr_graph_inputs) {
+  for (auto& v : inputs) {
+    curr_graph_inputs.insert(v);
+  }
+}
+
+void InsertQuantDeQuantHelper::extractAndRunWeightObserver(
+    Module& module,
+    Value* self,
+    Node* observer) {
+  Value* observed_weight = observer->output();
+  TORCH_CHECK(
+      isWeight(module, observed_weight),
+      "extractAndRunWeightObserver can only be called on weight tensors");
+  Graph* g = observer->owningGraph();
+  std::vector<Node*> weight_subgraph;
+  remap_values_in_observer_subgraph_.clear();
+
+  // If the graph was already visited, return list of relevant nodes directly.
+  // Multiple module instances can share the same graph code, so we don't need
+  // to re-run the extraction process.
+  if (graph_to_subgraph_nodes_.count(g)) {
+    weight_subgraph = graph_to_subgraph_nodes_[g];
+  } else {
+    // Extract the subgraph nodes.
+    weight_subgraph.push_back(observer);
+    Node* n = observer;
+    // Track values in the subgraph for which producer needs to be found.
+    std::unordered_set<Value*> subgraph_node_inputs;
+    updateCurrInputSet(observer->inputs().vec(), subgraph_node_inputs);
+    while (n && n != self->node()) {
```

Comment:
I think it's better to check for input values against self value

Reply:
we should also check for https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit13hitGraphInputEPNS0_5ValueE
you can move this function to Value class I think

Reply:
since we need to reject the following case:
```python
def forward(self, x, m):
       weight = m(x)
       return F.linear(x, weight)
```

Reply:
Not sure I follow, why would the case for node check vs input value check differ for the example above?

Reply:
value `m` in the example will have the same node as self, but it should be rejected, since in the end we are only going to pass in `self` as argument